servant/src/Servant/API/Authentication.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE GADTs              #-}
+{-# LANGUAGE KindSignatures     #-}
+{-# LANGUAGE PolyKinds          #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# OPTIONS_HADDOCK not-home    #-}
+module Servant.API.Authentication
+( AuthPolicy (..)
+, AuthProtect
+, AuthProtected (..)
+, BasicAuth (..)
+, JWTAuth (..)
+, OnMissing (..)
+, OnUnauthenticated (..)
+, SAuthPolicy (..)
+) where
+
+
+import           Data.ByteString (ByteString)
+import           Data.Typeable   (Typeable)
+import           GHC.TypeLits    (Symbol)
+import           Data.Text       (Text)
+
+-- | we can be either Strict or Lax.
+-- Strict: all handlers under 'AuthProtect' take a 'usr' argument.
+--         when auth fails, we call user-supplied handlers to respond.
+-- Lax: all handlers under 'AuthProtect' take a 'Maybe usr' argument.
+--      when auth fails, we call the handlers with 'Nothing'.
+data AuthPolicy = Strict | Lax
+
+-- | Singleton types for AuthPolicy
+data SAuthPolicy (p :: AuthPolicy) where
```

Comment:
I'm not sure we need the singletons. I *think* we really only need the type-level tags.